### ISSUE TITLE
Protobuf compiler generated code produces warnings in gcc 6.1 due to confusing indentation.

### ISSUE THREAD

**Issue by halyavin** [google/protobuf#1698]
The function MessageGenerator::GenerateIsInitialized in C++ generator (compiler/cpp/cpp_message.cc) inserts line break before instead of after if line:
```
  if (descriptor_->extension_range_count() > 0) {
    printer->Print(
      "\n"
      "if (!_extensions_.IsInitialized()) return false;");
  }
```

This causes final "return true;" at the end of generated function to be on the same line as this if statement. Such if statement produce confusing indentation warnings in GCC 6.1 since "return true;" looks like part of then branch with brackets missing. All other blocks in this function insert line breaks at the end of their generated code.

**Comment by xfxyjwf** [comment]
I think adding a "\n" at the end of the if statement would fix the issue?

**Comment by halyavin** [comment]
Yes, it would. I don't know what the leading line break is for though. According to blame, it was introduced before export to github and so I can't access its history. Also I don't know what is the policy of inserting empty lines in generated code. Maybe they are always inserted before this kind of if statements?

**Comment by onlyjob** [comment]
In GCC 6.2 they are no longer warnings. This problem seems to be causing FTBFS: #1797

**Comment by onlyjob** [comment]
See corresponding Debian bug: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=835266

**Comment by opoplawski** [comment]
Actually, it's still a warning, but certain tests in protobuf are compiled with -Werror.

**Comment by fenwanderer** [comment]
This appears to be fixed by 98835fb in v3.1.0.

**Comment by xfxyjwf** [comment]
@fenwanderer Thanks, yes, this is fixed in 3.1.0
